tools/comms.py

```diff
@@ -283,9 +291,12 @@ def eraseFlashSector(self, server_id, sector_nums):
 
     def programFlash(self, server_id, dest_addr, data):
         for i in range(0, len(data), COMM_SINGLE_PROGRAM_LENGTH):
-            success = self._programFlashLimitedLength(server_id, dest_addr + i, data[i:i+COMM_SINGLE_PROGRAM_LENGTH])
-            if not success:
-                return False
+            success = False
+            while not success:
+                try:
+                    success = self._programFlashLimitedLength(server_id, dest_addr + i, data[i:i+COMM_SINGLE_PROGRAM_LENGTH])
+                except (MalformedPacketError, ProtocolError) as e:
+                    print(f'Error: Retrying upload of segment {i} of {len(data)}')
```

Comment:
I see we are using f strings, which I like! Should we add a todo/issue to convert all of the other `.format` into f strings for consistency? (happy to help with that)

Reply:
Yes please! Any help would be much appreciated.